src/cls/rbd/cls_rbd.cc

```diff
@@ -2810,8 +2811,46 @@ int snapshot_set_limit(cls_method_context_t hctx, bufferlist *in,
   if (new_limit == UINT64_MAX) {
     CLS_LOG(20, "remove snapshot limit\n");
     rc = cls_cxx_map_remove_key(hctx, "snap_limit");
+    return rc;
+  }
+
+  //try to read header as v1 format
+  rc = snap_read_header(hctx, bl);
+
+  //if it fails, the image is v2 format
+  if (rc == -EINVAL) {
+    int max_read = RBD_MAX_KEYS_READ;
+    string last_read = RBD_SNAP_KEY_PREFIX;
+    bool more;
+
+    do {
+      set<string> keys;
+      rc = cls_cxx_map_get_keys(hctx, last_read, max_read, &keys, &more);
+      if (rc < 0) {
+        CLS_ERR("error retrieving snapshots: %s", cpp_strerror(rc).c_str());
+        return rc;
+      }
+      for (auto& key : keys) {
+        if (key.find(RBD_SNAP_KEY_PREFIX) != 0)
+          break;
+        snap_count++;
+      }
+      if (!keys.empty())
+        last_read = *(keys.rbegin());
+    } while (more);
+  } else {
```

Comment:
Nit: I would keep the v1 logic together -- so much this block up. I would also test for a failure to read the header:

```
if (rc < 0 && rc != -EINVAL) {
 ... error
} else if (rc >= 0) {
  .. v1 logic
} else {
  .. v2 logic
}
```